rust/datafusion/src/physical_plan/string_expressions.rs

```diff
@@ -115,6 +119,40 @@ where
     }
 }
 
+/// Returns number of characters in the string.
+/// character_length_i32('josé') = 4
+pub fn character_length_i32(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array: &GenericStringArray<i32> = args[0]
+        .as_any()
+        .downcast_ref::<GenericStringArray<i32>>()
+        .unwrap();
+
+    // first map is the iterator, second is for the `Option<_>`
+    let result = string_array
+        .iter()
+        .map(|x| x.map(|x: &str| x.graphemes(true).count() as i32))
+        .collect::<Int32Array>();
+
+    Ok(Arc::new(result) as ArrayRef)
+}
+
+/// Returns number of characters in the string.
+/// character_length_i64('josé') = 4
+pub fn character_length_i64(args: &[ArrayRef]) -> Result<ArrayRef> {
```

Comment:
I think that this code can be reduced to a single generic. Something like this (untested):

```rust
pub fn character_length<T: ArrowNativeType>(args: &[ArrayRef]) -> Result<ArrayRef> 
where 
T::Native: StringOffsetSizeTrait {
    let string_array: &GenericStringArray<T::Native> = args[0]
        .as_any()
        .downcast_ref::<GenericStringArray<T::Native>>()
        .unwrap();
    
     string_array
        .iter()
        .map(|x| x.map(|x: &str| T::Native::from_usize(x.graphemes(true).count())).unwrap())
        .collect()
}
```

Reply:
Thanks I will have a go. There are a couple more instances this generic from_usize could help.

Reply:
Wow. Impressive you can come up with this in your head. I have incorporated this code.